### ISSUE TITLE
Terraform assumes aws_appautoscaling_policy names are globally unique

### ISSUE THREAD

**Issue by hashibot** [terraform-providers/terraform-provider-aws#751]
_This issue was originally opened by @cbroglie as hashicorp/terraform#14302. It was migrated here as part of the [provider split](https://www.hashicorp.com/blog/upcoming-provider-changes-in-terraform-0-10/). The original body of the issue is below._

<hr>

AWS only requires the policy name to be unique given a resource id, service namespace, and scalable dimension, but Terraform assumes the policy names are globally unique. This makes the behavior of `aws_appautoscaling_policy` non-deterministic when the same policy name is reused for different resources.

### Terraform Version
v0.9.4

### Affected Resource(s)
Please list the resources as a list, for example:
- `aws_appautoscaling_policy`

### Terraform Configuration Files
Our configuration leverages a module named `ecs_autoscaling`, and its contents are here:
```hcl
variable "service_name" {}
variable "cluster_name" {}
variable "min" {}
variable "max" {}
variable "iam_role" {}

variable "cooldown" {
  default = 300
}

variable "metric_aggregation_type" {
  default = "Average"
}

variable "scale_up_adjustment" {
  default = 1
}

variable "scale_down_adjustment" {
  default = -1
}

resource "aws_appautoscaling_target" "ecs_target" {
  min_capacity = "${var.min}"
  max_capacity = "${var.max}"
  resource_id = "service/${var.cluster_name}/${var.service_name}"
  role_arn = "${var.iam_role}"
  scalable_dimension = "ecs:service:DesiredCount"
  service_namespace  = "ecs"
}

resource "aws_appautoscaling_policy" "ecs_policy_up" {
  depends_on = ["aws_appautoscaling_target.ecs_target"]

  adjustment_type = "ChangeInCapacity"
  cooldown = "${var.cooldown}"
  metric_aggregation_type = "${var.metric_aggregation_type}"
  name = "scale-up"
  resource_id = "service/${var.cluster_name}/${var.service_name}"
  scalable_dimension = "ecs:service:DesiredCount"
  service_namespace = "ecs"

  step_adjustment {
    metric_interval_lower_bound = 0
    scaling_adjustment = "${var.scale_up_adjustment}"
  }
}

resource "aws_appautoscaling_policy" "ecs_policy_down" {
  depends_on = ["aws_appautoscaling_target.ecs_target"]

  adjustment_type = "ChangeInCapacity"
  cooldown = "${var.cooldown}"
  metric_aggregation_type = "${var.metric_aggregation_type}"
  name = "scale-down"
  resource_id = "service/${var.cluster_name}/${var.service_name}"
  scalable_dimension = "ecs:service:DesiredCount"
  service_namespace = "ecs"

  step_adjustment {
    metric_interval_lower_bound = 0
    scaling_adjustment = "${var.scale_down_adjustment}"
  }
}

output "target_resource_id" {
  value = "${aws_appautoscaling_target.ecs_target.resource_id}"
}

output "scale_up_policy_arn" {
  value = "${aws_appautoscaling_policy.ecs_policy_up.arn}"
}

output "scale_down_policy_arn" {
  value = "${aws_appautoscaling_policy.ecs_policy_down.arn}"
}
```

Then the main configuration file leverages it like this:
```hcl
module "ecs_autoscaling_example1" {
  source  = "modules/ecs_autoscaling"
  service_name = "example1"
  cluster_name = "${aws_ecs_cluster.cluster.name}"
  min = "${var.min}"
  max = "${var.max}"
  iam_role = "${var.ecs_iam_role}"
}

module "ecs_autoscaling_example2" {
  source  = "modules/ecs_autoscaling"
  service_name = "example2"
  cluster_name = "${aws_ecs_cluster.cluster.name}"
  min = "${var.min}"
  max = "${var.max}"
  iam_role = "${var.ecs_iam_role}"
}
```

### Actual Behavior
Since Terraform assumes the names are globally unique, the behavior depends on the order of results from the call to `applicationautoscaling.DescribeScalingPolicies`.

### Steps to Reproduce
Execute `terraform apply` multiple times and observe that changes are made continuously.

### Important Factoids
The issue is [getAwsAppautoscalingPolicy](https://github.com/hashicorp/terraform/blob/b30ef0f58d6476dd021122471d7ae687c99e56bb/builtin/providers/aws/resource_aws_appautoscaling_policy.go#L289) only passes the policy name and service namespace parameters to `applicationautoscaling.DescribeScalingPolicies`, and uses the first result which matches the policy name. It should also pass the resource id and scalable dimension parameters, and assert that it gets at most 1 result back.

**Comment by bflad** [comment]
@cbroglie is this still an issue? I believe this might have been fixed in #1808 and released in provider version 1.1.0 (October 16, 2017).

**Comment by cbroglie** [comment]
Unfortunately I've switched companies, and am no longer setup to test.

But from looking back through my description above, I think #1808 is only a partial fix, as it still does not include the [resource id](https://github.com/aws/aws-sdk-go/blob/a201bf33b18ad4ab54344e4bc26b87eb6ad37b8e/service/applicationautoscaling/api.go#L3518).

**Comment by mksm** [comment]
This is still an issue. Having scaling policies with same name and attributes but for different `resource_id` causes terraform to keep wanting to update the resource's `resource_id`.

**Comment by bflad** [comment]
I have submitted a PR to address this: #3012

**Comment by bflad** [comment]
This has landed in master and will ship with [v1.7.1](https://github.com/terraform-providers/terraform-provider-aws/milestone/9) set to release later this week. Happy Terraform'ing! 🎉

**Comment by bflad** [comment]
This has been released in [terraform-provider-aws version 1.7.1](https://github.com/terraform-providers/terraform-provider-aws/blob/master/CHANGELOG.md#171-january-19-2018). Please see the [Terraform documentation on provider versioning](https://www.terraform.io/docs/configuration/providers.html#provider-versions) or reach out if you need any assistance upgrading.
